### PR TITLE
13 use hash in nodes

### PR DESCRIPTION
**Pull request by heidiklum**
hoping to close issue #13

### CODE REVIEW

**Comment by heidiklum** [comment]
instructions on how to run it are in:

https://github.com/bushidocodes/chord-grpc/issues/13

File: node.js [hunk 1]
```diff
@@ -1420,34 +1431,114 @@ async function migrateKeys() {}
...
- * --ip         - This node's IP Address'
- * --port       - This node's Port
+ * Takes the following mandatory flags
+ * --ip         - This node's IP Address
+ * --port       - This node's TCP Port
  *
  * --targetId   - The ID of a node in the cluster
  * --targetIp   - The IP of a node in the cluster
- * --targetPort - The Port of a node in the cluster
+ * --targetPort - The TCP Port of a node in the cluster
```
Comment (bushidocodes) [inline node.js:1440]: Good catch!

File: node.js [hunk 2]
```diff
@@ -24,6 +24,8 @@ const chord = grpc.loadPackageDefinition(packageDefinition).chord;
 const userMap = {};
 const HASH_BIT_LENGTH = 3;
 const CHECK_NODE_TIMEOUT_ms = 1000;
+const DEFAULT_HOST_NAME = "localhost";
```
Comment (bushidocodes) [inline node.js:27]: These are nice. What do you think about using hostName in place of ip throughout our codebase? If you like the idea, we should open an issue to refactor.

File: node.js [hunk 3]
```diff
@@ -1420,34 +1431,114 @@ async function migrateKeys() {}
...
-    args.targetIp !== null &&
-    args.targetPort !== null &&
-    args.targetId !== null
-  ) {
-    await join({ id: args.targetId, ip: args.targetIp, port: args.targetPort });
-  } else {
-    await join(null);
+  // compute identity parameters from arguments
+  _self.id = args.id ? args.id : null;
+  _self.ip = args.ip ? args.ip : DEFAULT_HOST_NAME;
```
Comment (bushidocodes) [inline node.js:1454]: Much cleaner!

File: utils.js [hunk 4]
```diff
@@ -71,7 +72,60 @@ function sha1(source) {
...
+ *
+ * @param {string} stringForHashing
+ * @param {number} hashBitLength
+ * @returns {number}
+ */
+async function computeIntegerHash(stringForHashing, hashBitLength) {
+  // enable debugging output
+  const DEBUGGING_LOCAL = false;
+
+  const MAX_JS_INT_BIT_LENGTH = 32;
```
Comment (bushidocodes) [inline utils.js:88]: Thank you for avoiding magic numbers!

**Review by bushidocodes — APPROVED** [review]
There were several collisions due to our tiny number of buckets, but I was able to get three nodes to join and then invoke the Data API against them, triggering both remote and local lookup and insertions. LGTM.

File: node.js (hunk 2 above)
Comment (heidiklum) [reply node.js:27]: Agreed. See issue #63

**Pull request merged**